addons/spreadsheet/static/src/data_sources/display_name_repository.js

```diff
@@ -0,0 +1,200 @@
+/** @odoo-module */
+
+import { Deferred } from "@web/core/utils/concurrency";
+import { LoadingDataError } from "../o_spreadsheet/errors";
+import BatchEndpoint, { Request } from "./server_data";
+
+/**
+ * @typedef PendingDisplayName
+ * @property {"PENDING"} state
+ * @property {Deferred<string|undefined>} deferred
+ *
+ * @typedef ErrorDisplayName
+ * @property {"ERROR"} state
+ * @property {Deferred<string|undefined>} deferred
+ * @property {Error} error
+ *
+ * @typedef CompletedDisplayName
+ * @property {"COMPLETED"} state
+ * @property {Deferred<string|undefined>} deferred
+ * @property {string|undefined} value
+ *
+ * @typedef {PendingDisplayName | ErrorDisplayName | CompletedDisplayName} DisplayNameResult
+ *
+ * @typedef {[number, string]} NameGetRPCResult
+ */
+
+/**
+ * This class is responsible for fetching the display names of records. It
+ * caches the display names of records that have already been fetched.
+ * It also provides a way to wait for the display name of a record to be
+ * fetched.
+ */
+export class DisplayNameRepository {
+    /**
+     *
+     * @param {import("@web/core/orm_service").ORM} orm
+     * @param {Object} params
+     * @param {function} params.whenDataIsFetched Callback to call when the
+     *  display name of a record is fetched.
+     */
+    constructor(orm, { whenDataIsFetched }) {
+        this.dataFetchedCallback = whenDataIsFetched;
+        /**
+         * Contains the display names of records. It's organized in the following way:
+         * {
+         *     "res.country": {
+         *         1: {
+         *              "value": "Belgium",
+         *              "deferred": Deferred<"Belgium">,
+         *     },
+         * }
+         */
+        /** @type {Object.<string, Object.<number, DisplayNameResult>>}*/
+        this._displayNames = {};
+        this._orm = orm;
+        this._endpoints = {};
+    }
+
+    /**
+     * Get the display name of the given record.
+     *
+     * @param {string} model
+     * @param {number} id
+     * @returns {Deferred<string|undefined>}
```

Comment:
could be typed as a Promise from the external point of view.
Does it really return undefined in some cases?